source/lac/cuda_vector.cc

```diff
@@ -325,136 +307,110 @@ namespace LinearAlgebra
-      Assert(dynamic_cast<const Vector<Number> *>(&V) != nullptr,
-             ExcVectorTypeNotCompatible());
-
-      // Downcast V. If fails, throw an exception.
-      const Vector<Number> &down_V = dynamic_cast<const Vector<Number> &>(V);
-      Assert(down_V.size() == this->size(),
+      Assert(V.size() == this->size(),
              ExcMessage(
                "Cannot add two vectors with different numbers of elements."));
 
-      // Check that casting will work.
-      Assert(dynamic_cast<const Vector<Number> *>(&W) != nullptr,
-             ExcVectorTypeNotCompatible());
-
-      // Downcast V. If fails, throw an exception.
-      const Vector<Number> &down_W = dynamic_cast<const Vector<Number> &>(W);
-      Assert(down_W.size() == this->size(),
+      Assert(W.size() == this->size(),
              ExcMessage(
                "Cannot add two vectors with different numbers of elements."));
 
       const int n_blocks = 1 + (n_elements - 1) / (chunk_size * block_size);
       kernel::add_aVbW<Number><<<dim3(n_blocks, 1), dim3(block_size)>>>(
-        val.get(), a, down_V.val.get(), b, down_W.val.get(), n_elements);
+        val.get(), a, V.val.get(), b, W.val.get(), n_elements);
       AssertCudaKernel();
     }
 
 
 
     template <typename Number>
     void
-    Vector<Number>::sadd(const Number                     s,
-                         const Number                     a,
-                         const VectorSpaceVector<Number> &V)
+    Vector<Number>::sadd(const Number          s,
+                         const Number          a,
+                         const Vector<Number> &V)
     {
       AssertIsFinite(s);
       AssertIsFinite(a);
 
-      // Check that casting will work.
-      Assert(dynamic_cast<const Vector<Number> *>(&V) != nullptr,
-             ExcVectorTypeNotCompatible());
-
-      // Downcast V. If fails, throw an exception.
-      const Vector<Number> &down_V = dynamic_cast<const Vector<Number> &>(V);
-      Assert(down_V.size() == this->size(),
+      Assert(V.size() == this->size(),
              ExcMessage(
                "Cannot add two vectors with different numbers of elements."));
 
       const int n_blocks = 1 + (n_elements - 1) / (chunk_size * block_size);
       kernel::sadd<Number><<<dim3(n_blocks, 1), dim3(block_size)>>>(
-        s, val.get(), a, down_V.val.get(), n_elements);
+        s, val.get(), a, V.val.get(), n_elements);
       AssertCudaKernel();
     }
 
 
 
     template <typename Number>
     void
-    Vector<Number>::scale(const VectorSpaceVector<Number> &scaling_factors)
+    Vector<Number>::scale(const Vector<Number> &scaling_factors)
     {
       // Check that casting will work.
       Assert(dynamic_cast<const Vector<Number> *>(&scaling_factors) != nullptr,
              ExcVectorTypeNotCompatible());
 
       // Downcast V. If fails, throw an exception.
-      const Vector<Number> &down_scaling_factors =
+      const Vector<Number> &scaling_factors =
```

Comment:
We don't need this line anymore

Reply:
Yes. I also learned in #15623 that there will likely be many errors still around the corner. I shouldn't have picked this class, since I can't compile it locally :-(